miscell fixes in README and images, and the embryo of a containers module

parmentelat opened the pull request:
* the example in the README now works fine
* fix broken `images.remove()`
* new `containers` submodule with `list_containers`, `inspect` and `kill` for starters

this builds on top of PR#18 that I closed before it was merged

parmentelat commented:
I have a couple questions regarding the doc that I found here <http://docs.podman.io/en/latest/_static/api.html>

* is this the right place to look ?
* if so what's the difference between the `containers` and `containers (compat)` tabs ?
* same for `images` *vs* `images (libpod)` ?

thanks !

parmentelat commented:
also I don't know how to handle the failed check, details don't help, any clue appreciated

rhatdan commented:
LGTM
Although could you fix your commit messages to be less then 80 characters, to remove the ugly line wrap.

@jwhonce @baude @TomSweeneyRedHat PTAL
The meta tests looks like something is broken in CI.

TomSweeneyRedHat commented:
LGTM
but would like a @jwhonce head nod

parmentelat commented:
I have rebased and cleaned up on top of the other PR
I have a few other things to propose but I'd rather not mix things up too much before I can get one batch through, please instruct if anything else needs to be tweaked

rhatdan commented:
Tests are failing.

parmentelat commented:
not quite sure what is wrong with the tests; simply running `pytest` shows 5/5 passing tests, it looks there's something with the coverage thingy that do not interpret properly ?

rhatdan commented:
Do you see the error if you run
make validate

parmentelat commented:
```
make validate
# I have no great ideas on how to install/check for git-validation
Validating vs commit ''
GIT_CHECK_EXCLUDE="./vendor:docs/make.bat" git-validation -run DCO,short-subject,dangling-whitespace -range 911121d43ca2707b2ff5aa7a22342aa74f5b2036..HEAD
/bin/sh: git-validation: command not found
make: *** [.gitvalidation] Error 127
```

parmentelat commented:
I think I get it: 
```
make unittest
coverage run -m unittest discover -s test/unit
......
----------------------------------------------------------------------
Ran 6 tests in 0.001s

OK
coverage report -m --skip-covered --fail-under=50 --omit=./test/*
Name                            Stmts   Miss  Cover   Missing
-------------------------------------------------------------
<snip>
-------------------------------------------------------------
TOTAL                             160     84    48%

1 file skipped due to complete coverage.
make: *** [unittest] Error 2
```

**`--fail-under=50`** and coverage is **`48%`**  
can you please confirm this is the problem ?

manics commented on podman/__init__.py:
```diff
@@ -1,4 +1,8 @@
 from podman.api_connection import ApiConnection
 
-__ALL__ = ["ApiConnection"]
+import podman.system
```
Is this right? `__ALL__` contains `system` whereas this imports it under `podman`

parmentelat replied:
not quite sure what the intention was with defining __ALL__ here in the first place, but first it should be `__all__` and not `__ALL__`, and second this only impacts `import * from podman` which is a dangerous practice anyways

otoh I was expecting for `import podman` to actually import stuff, hence this approach

jwhonce commented:
@parmentelat Please rebase on latest

parmentelat commented:
hi

I'm really unsure what to do with that PR; what I need is a basic toolset to list and manage containers; that's what I added on top of the code that I started with 6 months ago, and that's what I've been using in production since, and as you can see the needs are really scarce

I've taken a look at master;  there's still not even a `containers/` submodule in there, and trying to rebase suggests I'm going to have to redo the same work over again as the ground classes have changed

so all in all a little bit of context on a roadmap for this project, and why it is still so embryoid could help build motivation :)

rhatdan commented:
@parmentelat Most people are using docker-py to talk to the podman socket, at this point.  We are looking for volunteers to build out the full interaction with the podman socket for the full "libpod" api.  This is being mainly driven by community at this point, since the core developers have had different priorities.  We are trying to bring more focus on this effort now.  And hope to see it grow going forward.

If you can help to implement some of these missing features, it would be appreciated.

parmentelat commented:
thanks @rhatdan for the clarification, this is helpful

here you go, rebased and cleaned up

rhatdan commented:
@mwhahaha PTAL

mwhahaha commented on podman/containers/__init__.py:
```diff
@@ -0,0 +1,81 @@
+        path = api.join(path)
+
+    try:
+        response = api.request("DELETE", path)
```
api.delete(path, query)

mwhahaha commented on podman/images/__init__.py:
```diff
@@ -18,7 +18,7 @@ def inspect(api, name):
         return json.loads(str(response.read(), 'utf-8'))
     except errors.NotFoundError as e:
-        api.raise_not_found(e, e.response)
+        api.raise_image_not_found(e, e.response)
```
Yea this was a typo. I've caught this in #37 and it's fixed.

mwhahaha commented on podman/system/__init__.py:
```diff
@@ -28,6 +28,17 @@ def get_info(api):
+def info(api):
+    """Returns information on the system and libpod configuration"""
+    path = api.join("/info")
+    response = api.request("GET", path)
```
api.get("/info")

parmentelat commented:
all right so it sounds like i should wait until #37 gets merged, right ?

mwhahaha commented:
Yea I think it would be better. I wish this was gerrit instead of github so we could just manage code change dependencies better.

jwhonce commented:
@parmentelat Sorry, I stepped on some of your containers work in https://github.com/containers/podman-py/pull/38 . I stopped at list_containers() before abusing too much.  I did add the framework for container unit tests.

parmentelat commented:
well it feels like I'd better wait also for #38 to land then...

rhatdan commented:
@parmentelat Needs a rebase and #38 has landed.

rhatdan commented:
@parmentelat Are you still working on this?

parmentelat commented:
Hi there

ok, so, to be clear: what I had before - I mean 3 days ago when @rhatdan pinged me - was working for me, and I am not actively involved in this area of my system right now

I just took the time to rebase on top of your master branch; the thing is that version is broken for me; this is because of the `all_` argument to `list_containers`

I know `all` is a Python builtin and all, I know about pep008, but please note that it reads
> single_trailing_underscore_: used by convention to avoid conflicts with Python keyword, e.g.
and as you know `all` is just a builtin and not a keyword...

in any case that breaks my code; are you positive you want to go this road ?
I don't know how much you care about being in line with the docker API but their equivalent has no problem with using `all` as a keyword argument
https://docker-py.readthedocs.io/en/stable/containers.html#docker.models.containers.ContainerCollection.list
and it's my feeling that using `all_` instead is going to cause many people to waste their time trying `all` only to realize it's not the right argument name...

so for now I am reverting to using `all` so that my own code keeps on running

besides, I'd like to mention that `containers.list_containers()` is a bit redundant, and that `containers.list()` would do the trick just fine I expect

finally, I'd like to reiterate my comment about defining `__all__`  in the modules; it feels superfluous, what is the objective exactly here ? 

my 2 cents

parmentelat commented:
all right so, injecting this code into my deployment still breaks with this error

```
Nov 27 09:42:12 bash[3288865]: During handling of the above exception, another exception occurred:
Nov 27 09:42:12 bash[3288865]: Traceback (most recent call last):
Nov 27 09:42:12 bash[3288865]:   File "/usr/local/lib/python3.9/site-packages/nbhosting/courses/model_course.py", line 463, in image_hash
Nov 27 09:42:12 bash[3288865]:     return podman.images.inspect(podman_api, self.image)['Id']
Nov 27 09:42:12 bash[3288865]:   File "/root/podman-py/podman/images/__init__.py", line 22, in inspect
Nov 27 09:42:12 bash[3288865]:     api.raise_image_not_found(e, e.response)
Nov 27 09:42:12 bash[3288865]: AttributeError: 'ApiConnection' object has no attribute 'raise_image_not_found'
```

Looks like this method on ApiConnection has vanished in the meanwhile; is this intentional ?

reverting to my working code for now

mwhahaha commented:
yes it was renamed to raise_not_found and takes an exception type

rhatdan commented:
@parmentelat @mwhahaha Is this still being worked on? Or has it be usurped with changes in master?

mwhahaha commented:
This should be closed as I took these commits and used it to do #44
